datafusion/physical-plan/src/windows/mod.rs

```diff
@@ -174,20 +174,15 @@ fn create_built_in_window_expr(
     name: String,
     ignore_nulls: bool,
 ) -> Result<Arc<dyn BuiltInWindowFunctionExpr>> {
-    // need to get the types into an owned vec for some reason
-    let input_types: Vec<_> = args
-        .iter()
-        .map(|arg| arg.data_type(input_schema))
-        .collect::<Result<_>>()?;
+    // derive the output datatype from incoming schema
+    let out_data_type: &DataType = input_schema.field_with_name(&name)?.data_type();
 
-    // figure out the output type
-    let data_type = &fun.return_type(&input_types)?;
     Ok(match fun {
-        BuiltInWindowFunction::RowNumber => Arc::new(RowNumber::new(name, data_type)),
-        BuiltInWindowFunction::Rank => Arc::new(rank(name, data_type)),
-        BuiltInWindowFunction::DenseRank => Arc::new(dense_rank(name, data_type)),
-        BuiltInWindowFunction::PercentRank => Arc::new(percent_rank(name, data_type)),
-        BuiltInWindowFunction::CumeDist => Arc::new(cume_dist(name, data_type)),
+        BuiltInWindowFunction::RowNumber => Arc::new(RowNumber::new(name, out_data_type)),
```

Comment:
I find it strange in general that we have to pass a "output type" to `RowNumber` when it is (effectively) always the same.

Reply:
You are right @alamb, It is always the same for DataFusion, but when building the hybrid system, like something on top of the DF, it happens expected external output type its not the same as hardcoded in DF.

Imho, derivation from schema improves DF extensibility as it reduces coupling between physical and logical layers and the schema is the contract. Thus external systems on top of DF may safely use physical layer just providing the expected schema

Reply:
Reg to trait unification this is also a great idea to improve extensibility.
I'm thinking if can we move on with this PR, and migrate all window funcs to unified trait in follow up activity?

Reply:
> I'm thinking if can we move on with this PR, and migrate all window funcs to unified trait in follow up activity?

That seems fine to me (and the migration is already tracked, albiet at a high level) by another ticket